src/libraries/Native/Unix/System.Security.Cryptography.Native/opensslshim.c

```diff
@@ -150,11 +154,13 @@ void InitializeOpenSSLShim(void)
     fn##_ptr = (TYPEOF(fn))(dlsym(libssl, #fn));
 
 #define FALLBACK_FUNCTION(fn) \
-    if (!(fn##_ptr = (TYPEOF(fn))(dlsym(libssl, #fn)))) { fn##_ptr = (TYPEOF(fn))local_##fn; }
+    if (!(tmp_ptr = dlsym(libssl, #fn))) { tmp_ptr = (void*)local_##fn; } \
+    fn##_ptr = (TYPEOF(fn))tmp_ptr;
```

Comment:
The PR description describes race conditions due to concurrent initialization.  Do we need to use some kind of synchronization here?

Reply:
Synchronization and avoiding redundant work would be another solution.  But since dlsym is idempotent we can get back to a better state by avoiding writing intermediate nullptr values into the fn_ptr destinations, so that even if this code is running again it just (logically) does `foo_ptr = foo_ptr`, so there's never an observably bad state.

I expect to look into "avoiding redundant work" as a followup, just so that no other "this only runs once" assumptions are invalidated.

Reply:
I'm still not clear, though... what's to prevent the C compiler from reordering these global read/writes in fun ways?

Reply:
I was thinking, should the local `tmp_ptr` be volatile?